Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -272,47 +257,148 @@
-                        'mtm.elementVisibilityId': dom.getElementAttribute(entry.target, 'id'),
-                        'mtm.elementVisibilityClasses': dom.getElementClassNames(entry.target),
-                        'mtm.elementVisibilityText': TagManager.utils.trim(entry.target.innerText),
-                        'mtm.elementVisibilityNodeName': entry.target.nodeName,
-                        'mtm.elementVisibilityUrl': entry.target.href || dom.getElementAttribute(entry.target, 'href'),
-                    });
+        function isNodeEventTriggered(node) {
+            for (var j = 0; j < triggeredNodes.length; j++) {
+                if (node === triggeredNodes[j]) {
+                    return true;
+                }
+            }
 
-                    if (fireTriggerWhen === 'oncePage') {
-                        blockTrigger = true;
-                        TagManager.dom.bySelector(selectors).forEach(function (element) {
-                            observerIntersection.unobserve(element);
-                            triggeredNodes.push(element);
-                        });
-                        if (self.scrollIndex) {
-                            TagManager.window.offScroll(self.scrollIndex);
+            return false;
+        }
+
+        function setMutationObserver(triggerEvent) {
+            return function () {
+                if (observeDomChanges && isMutationObserverSupported) {
+                    notInDomNodes = getDynamicNodes(selectors);
+                    var targetNode = documentAlias.body;
+                    var config = {attributes: true, childList: true, subtree: true};
+                    observerMutation = new MutationObserver(function (mutationsList) {
+                        if (mutationObseverTimeout) {
+                            return;
                         }
-                    } else if (onlyOncePerElement) {
-                        observerIntersection.unobserve(entry.target);
-                        triggeredNodes.push(entry.target); // to avoid possible memory leaks as much as possible we add onceElement only when needed
+
+                        mutationObseverTimeout = true;
+                        setTimeout(function () {
+                            mutationObseverTimeout = false;
+                            mutationObserverCallback(mutationsList, triggerEvent);
+                            notInDomNodes = getDynamicNodes(selectors);
+                            if (notInDomNodes.length === 0) {
+                                observerMutation.disconnect();
+                            }
+                        }, 120);
+                    });
+
+                    observerMutation.observe(targetNode, config);
+                }
+            };
+        }
+
+        function mutationObserverCallback(mutationsList, triggerEvent) {
+            notInDomNodes.forEach(function (node) {
+                var element = documentAlias.querySelector(node);
```

Comment:
wouldn't we want to match all nodes in some cases? 
Even if `oncePerPage` or `oncePerElement` was selected then we'd still need to look over all nodes as one of them might be visible while others aren't? And it might not be the first one that is visible.

Reply:
Changed logic